### PR TITLE
Add support for hydrated aggregation results

### PR DESCRIPTION
**Pull request by alcaeus**
This PR builds on aggregation builder added in doctrine/mongodb#213 and adds hydrated aggregation results to ODM. It supersedes #1263 which will be closed for repository changes

Similar to `createQueryBuilder`, one now has `createAggregationBuilder` in DocumentManager and DocumentRepository classes. This will create an ODM aggregation builder that adds support for `references` and `includesReferenceTo` in match stages as well as type handling in those stages as well (type handling for other stages to be added later). The aggregation builder itself is extended to provide optional hydration for aggregation result documents via a new AggregationResultDocument type.

This new mapping type was necessary to handle the special cases for aggregation results:
- aggregation result documents may have complex IDs (think `$group` stages where the id might actually be an embedded document)
- aggregation result documents should not be managed by UnitOfWork
- aggregation result documents should not be persisted to the database

Note: hydrated aggregation will only be available for MongoDB 2.6 and later. MongoDB 2.2 and 2.4 do not support the cursor option for aggregation commands so no hydration results for those versions.

TODO list:
- [x] Add builder with hydration support
- [x] Ensure type handling and field name translation in match stage
- [x] Ensure type handling in other stages
- [x] Ensure field name translation in `$sort` and `$unwind` stages
- [x] Make `$lookup` stage work with referenced documents (TBD)
- [x] Ensure filters are applied to aggregation queries
- [x] Write tests for above cases
- [x] Write documentation

### CODE REVIEW

**Comment by speedytwenty** [comment]
Love following the progress on this! Very off topic, but I am very curious to know what tool you use to do the Todo checklist in a PR on github like you have . All of my searches for "github todo list" etc have turned up empty. Is the checklist integrated with something or is there a special syntax? I hate to take your time and impede on this PR but hoping you can point me in the right direction. Thanks!

**Comment by alcaeus** [comment]
This is part of GitHub Flavoured Markdown, scroll down to "Task Lists" in the [GFM documentation](https://guides.github.com/features/mastering-markdown/#GitHub-flavored-markdown).

**Comment by speedytwenty** [comment]
That's exactly what I've been looking for. Thank you very much!

**Comment by fabiorphp** [comment]
Any news about this PR? When the merge will be made?

**Comment by alcaeus** [comment]
I'm currently working on finalizing tests and writing documentation. No timeline though since I tend to not keep any deadlines I set myself. We're getting closer though :)

File: docs/en/reference/aggregation-builder.rst [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+Once you have mapped the document, use the ``hydrate()`` method to tell the
+aggregation builder about this document:
+
+.. code-block:: php
+
+    <?php
+
+    $builder = $dm->createAggregationBuilder('Orders');
+    $builder
+        ->hydrate('UserPurchases')
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: Wouldn't we want people to use FQCNs?

Comment (alcaeus) [inline docs/en/reference/aggregation-builder.rst]: Agreed - I should start doing it in tests and docs as well.

File: composer.json [hunk 2]
```diff
@@ -13,6 +13,12 @@
         { "name": "Maciej Malarz", "email": "malarzm@gmail.com" },
         { "name": "Andreas Braun", "email": "alcaeus@alcaeus.org" }
     ],
+    "repositories": [
+        {
+            "type": "vcs",
+            "url": "https://github.com/alcaeus/mongodb"
```
Comment (jmikola) [inline composer.json]: Is this a temporary change until https://github.com/doctrine/mongodb/pull/246 is merged and released?

File: docs/en/reference/aggregation-builder.rst [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                    ->month('purchaseDate')
+                    ->field('year')
+                    ->year('purchaseDate')
+            )
+            ->field('numPurchases')
+            ->sum(1)
+            ->field('amount')
+            ->sum('$amount');
+
+This aggregation would group all purchases by their month and year and providing
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: It took me a moment to realize exactly what this example was doing. Possible rephrasing:

> This aggregation would group all purchases by their month and year by projecting those values in an embedded object for the `_id` field. For example:

File: docs/en/reference/aggregation-builder.rst [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+.. code-block:: php
+
+    <?php
+
+    $builder = $dm->createAggregationBuilder('User');
+    $result = $builder->execute();
+
+If you instead want to look at the built aggregation pipeline, call the
+``getPipeline()`` method.
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: Should you clarify that this is the ``Builder::getPipeline()` method (I assume)?

File: docs/en/reference/aggregation-builder.rst [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+            ->field('amount')
+            ->sum('$amount');
+
+When you run the queries, all results will be returned as instances of the
+indicated document.
+
+.. note::
+
+    Query result documents can use all features regular documents can use: you
+    can map embedded documents, references and you can even use discriminators
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: Rephrase: "can map embedded documents, define references, and even use discriminators"

A comma after "references" helps readability for the list of possibilities.

File: docs/en/reference/aggregation-builder.rst [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    to get different result documents according to the aggregation result.
+    Unlike with regular documents, the identifier can be a complex object, e.g.
+    a reference or an embedded document.
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: Does ODM not allow `_id` to be mapped as an embedded documents for normal, persisted documents? I recall there being some PRs to handle non-scalar IDs but I'm not sure what became of it.

File: docs/en/reference/aggregation-builder.rst [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+builder object:
+
+.. code-block:: php
+
+    <?php
+
+    $builder = $dm->createAggregationBuilder(\Documents\Orders::class);
+    $builder
+        ->match()
+            ->field('purchaseDate')
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: Does ODM translate these strings if the PHP and MongoDB mappings differ? I can imagine that working for the first few pipeline operators but it becomes unreliable once we encounter the first operator that alters the document structure (e.g. `$project`, `$group`, `$unwind`).

File: composer.json (hunk 2 above)
Comment (alcaeus) [inline composer.json]: Yes. I was waiting for this to be done, then merge the corresponding PR and release 1.4, which will be a hard requirement for ODM 1.2.

File: docs/en/reference/aggregation-builder.rst (hunk 7 above)
Comment (alcaeus) [inline docs/en/reference/aggregation-builder.rst]: Yes it does, and yes it will get confusing. I've spent quite some time thinking about this and decided this was the best way to go. I should probably add this to the documentation as well, but in essence:
* if the field name given matches a mapped PHP field name in the document, it will be translated to the MongoDB field name
* additionally, if it is a reference or a DateTime value it will be prepared as is done in the query builder
* If a field name is reused (e.g. in a `$project` or `$group` stage) it will again be translated

Basically it boils down to: use the PHP names consistently and you'll be fine. Redefining existing fields should be fine in most cases (there may be some edge cases when doing `$unwind` but I haven't yet managed to think of them.

File: docs/en/reference/aggregation-builder.rst [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+            ->expression('$user')
+            ->field('numPurchases')
+            ->sum(1)
+            ->field('amount')
+            ->sum('$amount');
+
+Just like the query builder, the aggregation builder takes care of converting
+DateTime objects into MongoDate objects.
+
+Nesting objects
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: "Objects" sound general, maybe "expressions"?

File: docs/en/reference/aggregation-builder.rst [hunk 9]
```diff
@@ -0,0 +1,565 @@
...
+~~~~~~~~~~~~~~~
+
+You can create more complex objects by using the ``expr()`` method in the query
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: "more complex aggregatation stages"?

Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: Also should we refer to the builder as "aggregation builder" (and not "query builder")?

File: docs/en/reference/aggregation-builder.rst [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+                    ->field('year')
+                    ->year('purchaseDate')
+            )
+            ->field('numPurchases')
+            ->sum(1)
+            ->field('amount')
+            ->sum('$amount');
+
+This aggregation would group all purchases by their month and year by projecting
+those values in an embedded object for the ``id`` field. For example:
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: Not sure so will ask: are we "projecting in" or "projecting into" an embedded object?

File: docs/en/reference/aggregation-builder.rst [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+    {
+        _id: {
+            month: 1,
+            year: 2016
+        },
+        numPurchases: 1,
+        amount: 27.89
+    }
+
+Executing Queries
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: In first sentence we're talking about "Executing pipeline", should we rename header accordingly?

File: docs/en/reference/aggregation-builder.rst [hunk 12]
```diff
@@ -0,0 +1,565 @@
...
+~~~~~~~~~~~~~~~~~
+
+You can execute a pipeline using the ``execute()`` method. This will run the
+aggregation pipeline and return a cursor for you to iterate over the results:
+
+.. code-block:: php
+
+    <?php
+
+    $builder = $dm->createAggregationBuilder('User');
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: FQCN?

File: docs/en/reference/aggregation-builder.rst [hunk 13]
```diff
@@ -0,0 +1,565 @@
...
+        ->group()
+            ->field('id')
+            ->expression('$user')
+            ->field('numPurchases')
+            ->sum(1)
+            ->field('amount')
+            ->sum('$amount');
+
+When you run the queries, all results will be returned as instances of the
+indicated document.
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: How about "specified" document?

File: docs/en/reference/aggregation-builder.rst [hunk 14]
```diff
@@ -0,0 +1,565 @@
...
+
+.. note::
+
+    Query result documents can use all features regular documents can use: you
+    can map embedded documents, define references, and even use discriminators
+    to get different result documents according to the aggregation result.
+    Unlike with regular documents, the identifier can be a complex object, e.g.
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: I think I saw @jmikola asking the same question, but are complex objects disallowed as identifiers? I'm pretty sure I was using an object as an identifier (not mapped as embedded document though, just `\stdClass` or `array`)

File: docs/en/reference/aggregation-builder.rst [hunk 15]
```diff
@@ -0,0 +1,565 @@
...
+
+.. note::
+
+    The ``$lookup``, ``$sample`` and ``$indexStats`` stages were added in MongoDB 3.2.
+
+$project
+~~~~~~~~
+
+The ``$project`` stage lets you reshape the current document or define a completely
+new document:
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: "document" was just used, let's go with "one".

File: docs/en/reference/aggregation-builder.rst [hunk 16]
```diff
@@ -0,0 +1,565 @@
...
+        ->geoNear(120, 40)
+        ->spherical(true)
+        ->distanceField('distance')
+        // Convert radians to kilometers (use 3963.192 for miles)
+        ->distanceMultiplier(6378.137);
+
+.. note::
+
+    The ``$geoNear`` stage must be the first stage in the pipeline and the
+    collection must contain a single geospatial index. You must inclide the
```
Comment (malarzm) [inline docs/en/reference/aggregation-builder.rst]: there's a typo in "include"

File: lib/Doctrine/ODM/MongoDB/Aggregation/Builder.php [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+        parent::__construct($this->dm->getDocumentCollection($documentName));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function execute($options = array())
+    {
+        // Force cursor to be used
+        $options = array_merge($options, array('cursor' => true));
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Builder.php]: I think we wanted to use short array syntax for all new stuff?

File: lib/Doctrine/ODM/MongoDB/Aggregation/Builder.php [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+
+        $cursor = parent::execute($options);
+
+        return $this->prepareCursor($cursor);
+    }
+
+    /**
+     * @param string $className
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Builder.php]: Missing PHPDoc (as this method is not inherited)

File: lib/Doctrine/ODM/MongoDB/Aggregation/Builder.php [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+
+        return $query;
+    }
+
+    /**
+     * @param BaseCommandCursor $cursor
+     *
+     * @return CommandCursor
+     */
+    protected function prepareCursor(BaseCommandCursor $cursor)
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Builder.php]: Do we expect people to extend builder or can we keep this `private`?

File: lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Match.php [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+namespace Doctrine\ODM\MongoDB\Aggregation\Stage;
+
+use Doctrine\MongoDB\Aggregation\Stage as BaseStage;
+
+/**
+ * Fluent interface for building aggregation pipelines.
+ *
+ * @author alcaeus <alcaeus@alcaeus.org>
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Match.php]: You were removing all `@author` annotations in https://github.com/doctrine/mongodb-odm/pull/1326 :)

File: lib/Doctrine/ODM/MongoDB/CommandCursor.php [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+     *
+     * @return BaseCommandCursor
+     */
+    public function getBaseCursor()
+    {
+        return $this->commandCursor;
+    }
+
+    /**
+     * Rewind the cursor and return its first result.
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/CommandCursor.php]: Missing "s" at the end of verbs

File: lib/Doctrine/ODM/MongoDB/DocumentManager.php [hunk 22]
```diff
@@ -376,6 +376,17 @@ public function createQueryBuilder($documentName = null)
     }
 
     /**
+     * Create a new aggregation builder instance for a class.
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/DocumentManager.php]: CreateS

File: lib/Doctrine/ODM/MongoDB/DocumentRepository.php [hunk 23]
```diff
@@ -86,6 +86,16 @@ public function createQueryBuilder()
     }
 
     /**
+     * Create a new Aggregation\Builder instance that is prepopulated for this document name
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/DocumentRepository.php]: Here also "s" is missing and there's no "." at the end :)

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php [hunk 24]
```diff
@@ -38,9 +38,10 @@
 class AnnotationDriver extends AbstractAnnotationDriver
 {
     protected $entityAnnotationClasses = array(
-        ODM\Document::class         => 1,
-        ODM\MappedSuperclass::class => 2,
-        ODM\EmbeddedDocument::class => 3,
+        ODM\Document::class                  => 1,
+        ODM\MappedSuperclass::class          => 2,
+        ODM\EmbeddedDocument::class          => 3,
+        ODM\QueryResultDocument::class => 4,
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php]: I'd keep `=>` aligned now you've moved 3 previous ones :)

File: lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php [hunk 25]
```diff
@@ -117,14 +118,16 @@ public function loadMetadataForClass($className, ClassMetadata $class)
...
+            $class->isQueryResultDocument = true;
         }
         if (isset($documentAnnot->db)) {
             $class->setDatabase($documentAnnot->db);
         }
         if (isset($documentAnnot->collection)) {
             $class->setCollection($documentAnnot->collection);
         }
-        if (isset($documentAnnot->repositoryClass) && !$class->isEmbeddedDocument) {
+        if (isset($documentAnnot->repositoryClass) && !$class->isEmbeddedDocument && !$class->isQueryResultDocument) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Mapping/Driver/AnnotationDriver.php]: I think we can skip additional checks for embedded/query result as they're [in the setter anyway](https://github.com/doctrine/mongodb-odm/pull/1448/files#diff-85faac2da05627e5ef6ca0f86c69b3e1R558)

File: docs/en/reference/aggregation-builder.rst [hunk 26]
```diff
@@ -0,0 +1,563 @@
...
+    <?php
+
+    $builder = $dm->createAggregationBuilder(\Documents\Orders::class);
+    $builder
+        ->lookup('user')
+            ->alias('user')
+        ->unwind('$user');
+
+MongoDB will always return an array, even if the lookup only returned a single
+document. Thus, when looking up one-to-one referneces the result must be flattened
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: "referneces" typo

File: docs/en/reference/aggregation-builder.rst [hunk 27]
```diff
@@ -0,0 +1,563 @@
...
+using the ``$unwind`` operator.
+
+.. note::
+
+    Due to a limitation in MongoDB, the ``$lookup`` stage can only be used with
+    references that are stored as ID (see example above). References stored with a
+    DbRef object can't be used.
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: This limitation appears to be due to [SERVER-14466](https://jira.mongodb.org/browse/SERVER-14466).

Would it be worth linking to ODM's simple references here, since references still use DBRef by default?

File: docs/en/reference/aggregation-builder.rst [hunk 28]
```diff
@@ -0,0 +1,563 @@
...
+            year: 2016
+        },
+        purchaseDates: [
+            '2016-01-07',
+            '2016-03-10',
+            '2016-06-25'
+        ]
+    }
+
+To flatten the purchaseDates array, we would apply the following pipeline stage:
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: Should "purchaseDates" get monospace formatting as a field name?

File: docs/en/reference/aggregation-builder.rst [hunk 29]
```diff
@@ -0,0 +1,563 @@
...
+query options, allowing you to control the order and subset of results returned
+by the aggregation pipeline.
+
+$sample
+~~~~~~~
+
+.. note::
+
+    The ``$lookup`` stage was introduced in MongoDB 3.2. Using it on older servers
+    will result in errors.
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: This note looks like it was copy-pasted from an earlier section and should probably be deleted.

File: docs/en/reference/aggregation-builder.rst [hunk 30]
```diff
@@ -0,0 +1,563 @@
...
+            ->field('amount')
+            ->sum('$amount');
+
+$lookup
+~~~~~~~
+
+.. note::
+
+    The ``$lookup`` stage was introduced in MongoDB 3.2. Using it on older servers
+    will result in errors.
```
Comment (jmikola) [inline docs/en/reference/aggregation-builder.rst]: "in an error"?

**Comment by fabiorphp** [comment]
Any news about this PR? When the tag will be made?

**Comment by alcaeus** [comment]
The initial idea was to release 1.2 along a bunch of deprecations required for 2.0. Since work on 2.0 is currently on hold due to time issues on my end, we'll revisit the timeline and try to release 1.2 soon. I'll revisit the issue tracker to triage issues that need to go into 1.2 and report back.
